paddle/fluid/operators/unsqueeze_op.cc

```diff
@@ -0,0 +1,191 @@
+
+    const auto &axes = ctx->Attrs().Get<std::vector<int>>("axes");
+    PADDLE_ENFORCE(!axes.empty(),
+                   "The unsqueeze axes information must be set by Attr(axes).");
+
+    const auto &x_dims = ctx->GetInputDim("X");
+    // Validity Check: input tensor dims (<6).
+    PADDLE_ENFORCE(static_cast<int>(x_dims.size()) <= 6,
+                   "Invalid dimensions, dynamic dimensions should within "
+                   "[1, 6] dimensions (Eigen limit).");
+    // Validity Check: the range of unsqueeze aixs.
+    for (int axis : axes) {
+      PADDLE_ENFORCE(axis < 6,
+                     "Invalid dimensions, input axis should within "
+                     "[1, 6] dimensions (Eigen limit).");
+    }
+
+    auto out_dims = GetOutputShape(axes, x_dims);
+    ctx->SetOutputDim("Out", out_dims);
+  }
+
+  static framework::DDim GetOutputShape(const std::vector<int> unsqz_dims,
+                                        const framework::DDim &in_dims) {
+    unsigned int unsqz_mask = 0;
+    unsigned int front = 0, back = 0;
+    int output_dims_size = in_dims.size();
+
+    // Simulate insert by bit calc.
+    for (int axis : unsqz_dims) {
+      int cur = axis < 0 ? axis + output_dims_size + 1 : axis;
+      // Vaildity Check: the axis bound
+      PADDLE_ENFORCE(
+          cur >= 0 && cur <= output_dims_size,
+          "The unsqueeze dims must be within range of current rank.");
+      // Save the front part.
+      front = unsqz_mask & ((1 << cur) - 1);
+      // Move the back part.
+      back = unsqz_mask & ~((1 << cur) - 1);
+      back <<= 1;
+      // Merge two part.
+      back |= (1 << cur);
+      unsqz_mask = front | back;
+      // Add the output size.
+      output_dims_size++;
+      // Validity Check: rank range.
+      PADDLE_ENFORCE(output_dims_size <= 6,
+                     "The output tensor's rank should be less than 6.");
+    }
+
+    // Make output shape
+    std::vector<int64_t> output_shape(output_dims_size, 0);
+    for (int in_idx = 0, out_idx = 0; out_idx < output_dims_size; ++out_idx) {
+      if ((unsqz_mask & (1 << out_idx)) == 0) {
+        output_shape[out_idx] = in_dims[in_idx++];
+      } else {
+        output_shape[out_idx] = 1;
+      }
+    }
+
+    return framework::make_ddim(output_shape);
+  }
+};
+
+class UnsqueezeOp : public framework::OperatorBase {
+ public:
+  UnsqueezeOp(const std::string &type, const framework::VariableNameMap &inputs,
+              const framework::VariableNameMap &outputs,
+              const framework::AttributeMap &attrs)
+      : OperatorBase(type, inputs, outputs, attrs) {}
+
+ private:
+  void RunImpl(const framework::Scope &scope,
+               const platform::Place &place) const override {
+    auto &axes = Attr<std::vector<int>>("axes");
+    auto x_dims = scope.FindVar(Input("X"))->Get<framework::LoDTensor>().dims();
+    auto out_dims = UnsqueezeOpInferShape::GetOutputShape(axes, x_dims);
```

Comment:
You have set the output's dim in `UnsqueezeOpInferShape ` stage, so you can get the out_dims directly here.

Reply:
get, thanks

Reply:
It may not work here. InferShape is invoked in op->RunImpl implemented based OperatorWithKernel, but isn't invoked in op->Run based OperatorBase, the op->RunImpl should implement by ourself, so this place should calculate by function.

Reply:
I see, thanks!